lib/plugins/stackoverflow/post.dart

```diff
@@ -32,11 +35,86 @@ class StackOverflowPostOwner with _$StackOverflowPostOwner {
       _$StackOverflowPostOwnerFromJson(json);
 }
 
-String postToDiscordMarkdown(StackOverflowPost post) {
-  return '# ${post.title}\n'
+Future<String> postToDiscordMarkdown(
+  StackOverflowPost post,
+  Environment env,
+) async {
+  String header = '# ${post.title}\n'
       '[${post.owner.displayName}](<${post.owner.link}>) - '
       '${timeago.format(DateTime.fromMillisecondsSinceEpoch(post.creationDate * 1000))} | '
       '${post.score} votes | '
-      '[Source](<${post.link}>)\n\n'
-      '${post.bodyMarkdown}';
+      '[Source](<${post.link}>)\n\n';
+
+  String bodyMarkdown = env.pastebinApiKey == null
+      ? post.bodyMarkdown
+      : await shortenWithPastebin(
+          post.bodyMarkdown,
+          PastebinClient(env.pastebinApiKey!),
+          kMaxMessageLength - header.length,
+        );
+
+  String result = '$header$bodyMarkdown';
+
+  if (result.length > kMaxMessageLength) {
+    return '${result.substring(0, kMaxMessageLength - 3)}...';
+  }
+
+  return result;
+}
+
+// Matches code blocks fenced with ```code``` in group 1 and code blocks
+// indented by four spaces in group 2 (SO seems to use the latter quite often).
+final codeblockPattern = RegExp(
+  r'(?:```(?:dart|)\n((?:.*\n)+?)```)|(?:( {4}.+\n(?:(?: {4}.+\n|\n)*) {4}.+)\n)',
+  caseSensitive: false,
+);
+
+Future<String> shortenWithPastebin(
+  String content,
+  PastebinClient client,
+  int maxLength,
+) async {
+  content = content.replaceAll('\r\n', '\n');
+
+  final allMatches = codeblockPattern.allMatches(content);
+  // (start, length, content)
+  final blocks = List.of(allMatches.map((m) => (
+        m.start,
+        m.end - m.start,
+        m.group(1) ?? m.group(2)!,
+      )))
+    // Ignore code blocks less than 200 characters long.
+    ..removeWhere((element) => element.$2 < 200);
+
+  for (final block in blocks) {
+    // Cutting codeblocks that will be cut anyway has no effect
+    if (block.$1 > maxLength) break;
+    // We've managed to cut out enough blocks to make the content fit
+    if (content.length <= maxLength) break;
```

Comment:
does this mean we would leave in trailing large code blocks if they fit, just by chance, after we've trimmed all the other ones? while that does seem a bit unlikely, it also sounds undesirable.

Reply:
Yes, but they're cropped out by postToDiscordMarkdown cutting content to kMaxMessageLength. If we didn't do this, we'd be doing extra work (and performing extra API requests) for nothing.

Reply:
Oh, I realise I misread the comment. Yes, we do leave them uncropped.

I thought that'd be best as it saves us a few pastebin API requests. Originally I'd written the code so it crops out code blocks from largest to smallest until the limit is reached, but that also uses extra unneeded api requests.

Reply:
okay. I think we get a shit ton of request limit with the API key, but I dont really have a strong opinion on this. 
we can leave it that way then.